src/libopensc/card-iasecc.c

```diff
@@ -612,12 +612,17 @@ iasecc_init_mi(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
-    resp_len = sizeof(resp);
-    rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
-    LOG_TEST_RET(ctx, rv, "Could not select MI's AID");
+	resp_len = sizeof(resp);
+	rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
 
 	rv = iasecc_mi_match(card);
-    LOG_TEST_RET(ctx, rv, "Could not match MI's AID");
+
+	if(rv) {
+		card->type = SC_CARD_TYPE_IASECC_MI2;
+		rv = iasecc_init_sagem(card);
```

Comment:
Could you please unify the usage of `iasecc_init_amos`, `iasecc_init_sagem` and your card? Currently, they all do the same, which should be generalized under a different name.

I think the driver is not very consistently written; `iasecc_init_oberthur` and `iasecc_init_gemalto` also shares most of the code...